rosserial_arduino/nodes/serial_node.py

```diff
@@ -0,0 +1,100 @@
+#    with the distribution.
+#  * Neither the name of Willow Garage, Inc. nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import rospy
+from rosserial_arduino import SerialClient
+from rosserial_python import RosSerialServer
+from serial import SerialException
+from time import sleep
+import multiprocessing
+
+import sys
+
+if __name__=="__main__":
+
+    rospy.init_node("serial_node")
+    rospy.loginfo("ROS Serial Python Node")
+
+    port_name = rospy.get_param('~port','/dev/ttyUSB0')
+    baud = int(rospy.get_param('~baud','57600'))
+
+    # Number of seconds of sync failure after which Arduino is auto-reset.
+    # 0 = no timeout, auto-reset disabled
+    auto_reset_timeout = int(rospy.get_param('~auto_reset_timeout','0'))
+
+    # for systems where pyserial yields errors in the fcntl.ioctl(self.fd, TIOCMBIS, \
+    # TIOCM_DTR_str) line, which causes an IOError, when using simulated port
+    fix_pyserial_for_test = rospy.get_param('~fix_pyserial_for_test', False)
+
+    # TODO: should these really be global?
+    tcp_portnum = int(rospy.get_param('/rosserial_embeddedlinux/tcp_port', '11411'))
+    fork_server = rospy.get_param('/rosserial_embeddedlinux/fork_server', False)
+
+    # TODO: do we really want command line params in addition to parameter server params?
+    sys.argv = rospy.myargv(argv=sys.argv)
+    if len(sys.argv) >= 2 :
+        port_name  = sys.argv[1]
+    if len(sys.argv) == 3 :
+        tcp_portnum = int(sys.argv[2])
+
+    if port_name == "tcp" :
+        server = RosSerialServer(tcp_portnum, fork_server)
+        rospy.loginfo("Waiting for socket connections on port %d" % tcp_portnum)
+        try:
+            server.listen()
+        except KeyboardInterrupt:
+            rospy.loginfo("got keyboard interrupt")
+        finally:
+            rospy.loginfo("Shutting down")
+            for process in multiprocessing.active_children():
+                rospy.loginfo("Shutting down process %r", process)
+                process.terminate()
+                process.join()
+            rospy.loginfo("All done")
```

Comment:
Are there arduinos using the TCP connection? If not, let's remove this section and the params above.

Reply:
Done.